From 5ac3a6770ce45c1ee05ddc6c84a236df72785287 Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 26 Jun 2014 15:54:27 -0400
Subject: [net] sunrpc/xprtrdma: Allocate missing pagelist

Message-id: <1403798077-28799-69-git-send-email-steved@redhat.com>
Patchwork-id: 84918
O-Subject: [RHEL7.1 PATCH 68/78] xprtrdma: Allocate missing pagelist
Bugzilla: 1113248
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>

From: Shirley Ma <shirley.ma@oracle.com>

GETACL relies on transport layer to alloc memory for reply buffer.
However xprtrdma assumes that the reply buffer (pagelist) has been
pre-allocated in upper layer. This problem was reported by IOL OFA lab
test on PPC.

Signed-off-by: Shirley Ma <shirley.ma@oracle.com>
Reviewed-by: Chuck Lever <chuck.lever@oracle.com>
Tested-by: Edward Mossman <emossman@iol.unh.edu>
Signed-off-by: Anna Schumaker <Anna.Schumaker@Netapp.com>
Upstream Id: 196c69989d84ab902bbe545f7bd8ce78ee74dac4
Bugzilla: 1113248
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/sunrpc/xprtrdma/rpc_rdma.c b/net/sunrpc/xprtrdma/rpc_rdma.c
index 436d229..dc4a826 100644
--- a/net/sunrpc/xprtrdma/rpc_rdma.c
+++ b/net/sunrpc/xprtrdma/rpc_rdma.c
@@ -99,6 +99,12 @@ rpcrdma_convert_iovs(struct xdr_buf *xdrbuf, unsigned int pos,
  page_base = xdrbuf->page_base & ~PAGE_MASK;
  p = 0;
  while (len && n < nsegs) {
+  if (!ppages[p]) {
+   /* alloc the pagelist for receiving buffer */
+   ppages[p] = alloc_page(GFP_ATOMIC);
+   if (!ppages[p])
+    return 0;
+  }
   seg[n].mr_page = ppages[p];
   seg[n].mr_offset = (void *)(unsigned long) page_base;
   seg[n].mr_len = min_t(u32, PAGE_SIZE - page_base, len);
-- 
1.7.1